examples credentials ocmconfig (#323)
open-component-model/ocm-website: content/docs/examples/creds-in-ocmconfig.md

A `consumer` is something the credentials are required for. For example, if you need to configure credentials that are required to log in to an OCI registry, one could say that the registry will be consuming these credentials, i.e. the registry is a credentials consumer. That is why it is configured under `consumers`.

### Re-use credentials configured for Docker

This `.ocmconfig` file will tell OCM CLI to use configuration from Docker's `config.json`.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    repositories:
      - repository:
          type: DockerConfig/v1
          dockerConfigFile: "~/.docker/config.json"
          propagateConsumerIdentity: true
```

### Accessing OCI registries

#### HTTPS and path

> To access artifacts in `https://ghcr.io/mandelsoft`:

Reviewer: ```suggestion
To access artifacts in `https://ghcr.io/open-component-model`:
```

* Omit the protocol
* The path behind the top-level domain hast to be specified in a separate `pathprefix` field

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          hostname: ghcr.io

>           pathprefix: mandelsoft

Reviewer: ```suggestion
          pathprefix: open-component-model
```

        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-token
```

#### HTTP, port number, empty path

To access artifacts in `http://127.0.0.1:5001`:
* Omit the protocol
* Omit the port number
* As the URL has no path behind the port number, `pathprefix` element can be removed or left empty

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          hostname: 127.0.0.1
        credentials:
          - type: Credentials
            properties:
              username: admin
              password: admin
```

### Accessing Helm chart repositories

Similar to OCI registries, but use `HelmChartRepository` as identity type.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: HelmChartRepository
          hostname: ghcr.io

>           pathprefix: mandelsoft

Reviewer: ```suggestion
          pathprefix: open-component-model
```

* `token` is a personal access token generated in GitHub Developer Settings

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: Github
          hostname: my.github.enterprise
          pathprefix: my-org/my-repo
        credentials:
          - type: Credentials
            properties:
              token: ghp_my_personal_access_token
```

### Accessing several systems

It is, of course, possible to configure credentials for several systems in the same `.ocmconfig` file. To do that, you can combine as many repositories and consumers as you need.

The example below instructs OCM CLI to look for credentials in Docker's `config.json`, and in addition specifies dedicated credentials for an OCI registry and a GitHub repository.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    repositories:
      - repository:
          type: DockerConfig/v1
          dockerConfigFile: "~/.docker/config.json"
          propagateConsumerIdentity: true
    consumers:
      - identity:
          type: OCIRegistry
          hostname: ghcr.io

>           pathprefix: mandelsoft

Reviewer: ```suggestion
          pathprefix: open-component-model
```

---
title: "Credentials in ocmconfig file"
description: ""
summary: ""
date: 2024-09-04T13:54:01+02:00
lastmod: 2024-09-04T13:54:01+02:00
draft: false
weight: 999
toc: true
---

## Overview

The [OCM command line client](https://github.com/open-component-model/ocm/blob/main/docs/reference/ocm.md) can be configured by supplying it with a [configuration file](https://github.com/open-component-model/ocm/blob/main/docs/reference/ocm_configfile.md). By default, the CLI looks for configuration in `$HOME/.ocmconfig`, if it exists.

The configuration file can be used in particular to specify the credentials, which are required for the CLI to be able to access the artifact repositories, referenced in CLI commands.

## Examples

This page contains basic examples of credentials configuration for a few most common artifact repository types. The examples below are complete `.ocmconfig` files, not snippets.

For comprehensive documentation on the credentials topic, including usage of certificates or HashiCorp Vault, execute the command `ocm credential-handling`.

### Repositories and Consumers

> In the examples below some configuration is located under `configurations[0]`.`repositories`, and some other under `configurations[0]`.`consumers`. This chanpter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.

Reviewer 2: ```suggestion
In the examples below some configuration is located under `configurations[0]`.`repositories`, and some other under `configurations[0]`.`consumers`. This chapter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.

```

> In the examples below some configuration is located under `configurations[0]`.`repositories`, and some other under `configurations[0]`.`consumers`. This chanpter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.

Reviewer 3: ```suggestion
In the examples below some configuration is located under `configurations[0]`.`repositories`, and some other under `configurations[0]`.`consumers`. This chapter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.

```

> In the examples below some configuration is located under `configurations[0]`.`repositories`, and some other under `configurations[0]`.`consumers`. This chanpter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.

Reviewer 3: ```suggestion
In the examples below some configuration is located under `configurations.repositories`, and some other under `configurations.consumers`. This chanpter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.

```

Although the notation `[0]` is technically correct, it can disturb the reading-flow. I would omit them

Author: Speaking of reading-flow, I'm one of the people, whose reading-flow would immediately be disturbed by incorrect technical references. My next questions would then be, what else might be wrong with the page, and if I can trust this technical documentation at all? Therefore I'd keep the index.

In this context, `repository` is a place, where credentials can be stored, i.e. it is a credentials repository. For example, Docker's `config.json` can store multiple credentials, and in that sense the file serves as a repository that can store and provide credentials. That is why its location is configured under `repositories`. Other examples of credentials repositories can be the NPM's `.npmrc` file or a HashiCorp Vault instance.

A `consumer` is something the credentials are required for. For example, if you need to configure credentials that are required to log in to an OCI registry, one could say that the registry will be consuming these credentials, i.e. the registry is a credentials consumer. That is why it is configured under `consumers`.

### Re-use credentials configured for Docker

> This `.ocmconfig` file will tell OCM CLI to use credentials configuration from Docker's `config.json` file.

Reviewer 2: ```suggestion
This `.ocmconfig` file will tell the OCM CLI to use credentials configuration from Docker's `config.json` file.
```

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    repositories:
      - repository:
          type: DockerConfig/v1
          dockerConfigFile: "~/.docker/config.json"

>           propagateConsumerIdentity: true

Reviewer 2: After having a quick glance at our library code, it appears as if our `propagateConsumerIdentity` behaviour is quite inconsistent. 
Every credential repository spec may contain this property. But while e.g. _dockerconfig_ defaults `propagateConsumerIdentity` to `false` if it is not specified, _npm_  defaults `propagateConsumerIdentity` to `true`, if it is not specified. As to my understanding, this property has to be set to true in order to be actually useful for _command line usage_. That's why it is required here!
We should align this for all credential repositories and I'd argue we should default to true, since command line usage is the primary use case for the config file. Furthermore, I think a common ocm cli user should not have to bother trying to understand this property.

Reviewer 2: I was wrong, it is defaulted for dockerconfig as well!

Author: What is your suggestion? Uwe's documentation has several examples with `propagateConsumerIdentity: true`, e.g. [here](https://github.com/open-component-model/ocm/blob/main/docs/reference/ocm_configfile.md). I'd prefer to keep the line, to have it consistent. Second, the purpose of this page is to provide a working example, and in this form the file was working for my use cases.

Author: Suggestion to remove `propagateConsumerIdentity: true` accepted

>           propagateConsumerIdentity: true

Reviewer 2: ```suggestion

```

Reviewer 2: its defaulted to true and can therefore be omitted.

```

### Re-use credentials configured for npm

This `.ocmconfig` file will tell OCM CLI to use credentials configuration from npm's `.npmrc` file.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    repositories:
      - repository:
          type: NPMConfig/v1

>           npmrcFile: '~/.npmrc'

Reviewer 2: Here `propagateConsumerIdentity` can be omitted, since its defaulted to `true`!

Author: The purpose of this page is to provide a working example, and we know that  in this form the file was working for @hilmarf. I'd suggest keeping the line `propagateConsumerIdentity: true`, as it does not make anything worse. And please also see my reply regarding `DockerConfig `.

```

### Accessing OCI registries

#### HTTPS and path

To access artifacts in `https://ghcr.io/open-component-model`:

> * Omit the protocol

Reviewer 2: Weird to write it like this here. For me as a reader, this gives the impression that the protocol cannot be included in the consumer id.
My suggestion would be to either just write:

To access artifacts in `ghcr.io/open-component-model`: ...
So, omit the scheme (or protocol) and also omit this whole list point.

or keep `https://ghcr.io/open-component-model`, but then also include https in the consumer id below.
```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          scheme: https
          hostname: ghcr.io
          pathprefix: open-component-model
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-token
```

Author: I vote for the second option. I'll re-work.

Author: OK, what about the following?

#### HTTPS and path

To access artifacts in `https://ghcr.io/open-component-model`:
* The different parts of the URL have to be specified in separate fields: `scheme`, `hostname` and `pathprefix`
* The fields `scheme` and `pathprefix` are optional. If not specified, the OCM CLI will use the credentials for all schemes and paths on that host
* The `password` is the user's basic authentication password. Some OCI registries allow to generate user access tokens, which can also be used for basic authentication.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          scheme: https
          hostname: ghcr.io
          pathprefix: open-component-model
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-token
```

Author: Suggestion implemented

> * The path behind the top-level domain hast to be specified in a separate `pathprefix` field

Reviewer 2: ```suggestion
* The path behind the top-level domain has to be specified in a separate `pathprefix` field
```

* The `password` is the user's basic authentication password. Some OCI registries allow to generate user access tokens, which can also be used for basic authentication.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          hostname: ghcr.io
          pathprefix: open-component-model
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-token
```

#### HTTP, port number, empty path

To access artifacts in `http://127.0.0.1:5001`:

> * Omit the protocol

Reviewer 2: Essentially the same critique as above. What's the rationale behind omitting it here?
What if the goes ahead, starts up another oci registry on port 5000, resulting in the address `http://127.0.0.1:5000`?

Author: What about the following?

#### HTTP, port number, empty path

To access artifacts in `http://127.0.0.1:5001`:
* The field `port` is optional. If not specified, the OCM CLI will use the credentials for all ports on that host
* As the URL has no path behind the port number, `pathprefix` element can be removed or left empty

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          scheme: http
          hostname: 127.0.0.1
          port: 5001
        credentials:
          - type: Credentials
            properties:
              username: admin
              password: admin
```

Author: Suggestion implemented

---

> title: "Credentials in ocmconfig file"

Reviewer 4: ```suggestion
title: "Credentials in an ocmconfig file"

```

description: ""
summary: ""
date: 2024-09-04T13:54:01+02:00
lastmod: 2024-09-04T13:54:01+02:00
draft: false
weight: 999
toc: true
---

## Overview

The [OCM command line client](https://github.com/open-component-model/ocm/blob/main/docs/reference/ocm.md) can be configured by supplying it with a [configuration file](https://github.com/open-component-model/ocm/blob/main/docs/reference/ocm_configfile.md). By default, the CLI looks for configuration in `$HOME/.ocmconfig`, if it exists.

> The configuration file can be used in particular to specify the credentials, which are required for the CLI to be able to access the artifact repositories, referenced in CLI commands.

Reviewer 4: ```suggestion
The configuration file can be used in particular to specify the credentials, which are required for the CLI to be able to access the artifact repositories referenced in CLI commands.

```

## Examples

This page contains basic examples of credentials configuration for a few most common artifact repository types. The examples below are complete `.ocmconfig` files, not snippets.

For comprehensive documentation on the credentials topic, including usage of certificates or HashiCorp Vault, execute the command `ocm credential-handling`.

### Repositories and Consumers

> In the examples below some configuration is located under `configurations[0]`.`repositories`, and some other under `configurations[0]`.`consumers`. This chapter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.

Reviewer 4: ```suggestion
In the examples below, some configuration is located under `configurations[0]`.`repositories`, and some other under `configurations[0]`.`consumers`. This chapter explains the difference between `repositories` and `consumers`, which is potentially not as obvious as one could think.
```

> 
> In this context, `repository` is a place, where credentials can be stored, i.e. it is a credentials repository. For example, Docker's `config.json` can store multiple credentials, and in that sense the file serves as a repository that can store and provide credentials. That is why its location is configured under `repositories`. Other examples of credentials repositories can be the NPM's `.npmrc` file or a HashiCorp Vault instance.

Reviewer 4: ```suggestion
In this context, `repository` is a place where credentials can be stored, i.e., it is a credentials repository. For example, Docker's `config.json` can store multiple credentials, and in that sense the file serves as a repository that can store and provide credentials. That is why its location is configured under `repositories`. Other examples of credentials repositories can be the NPM's `.npmrc` file or a HashiCorp Vault instance.

```

> A `consumer` is something the credentials are required for. For example, if you need to configure credentials that are required to log in to an OCI registry, one could say that the registry will be consuming these credentials, i.e. the registry is a credentials consumer. That is why it is configured under `consumers`.

Reviewer 4: ```suggestion
A `consumer` is something the credentials are required for. For example, if you need to configure credentials that are required to log in to an OCI registry, one could say that the registry will be consuming these credentials, i.e., the registry is a credentials consumer. That is why it is configured under `consumers`.

```

> ### Re-use credentials configured for Docker

Reviewer 4: ```suggestion
### Reuse Credentials Configured for Docker

```

This `.ocmconfig` file will tell the OCM CLI to use credentials configuration from Docker's `config.json` file.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    repositories:
      - repository:
          type: DockerConfig/v1
          dockerConfigFile: "~/.docker/config.json"
```

> ### Re-use credentials configured for npm

Reviewer 4: ```suggestion
### Reuse Credentials Configured for npm

```

This `.ocmconfig` file will tell OCM CLI to use credentials configuration from npm's `.npmrc` file.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    repositories:
      - repository:
          type: NPMConfig/v1
          npmrcFile: '~/.npmrc'
```

> ### Accessing OCI registries

Reviewer 4: ```suggestion
### Accessing OCI Registries

```

> #### HTTPS and path

Reviewer 4: ```suggestion
#### HTTPS and Path

```

To access artifacts in `https://ghcr.io/open-component-model`:

> * The different parts of the URL have to be specified in separate fields: `scheme`, `hostname` and `pathprefix`
> * The fields `scheme` and `pathprefix` are optional. If not specified, the OCM CLI will use the credentials for all schemes and paths on that host
> * The `password` is the user's basic authentication password. Some OCI registries allow to generate user access tokens, which can also be used for basic authentication.

Reviewer 4: ```suggestion
* The different parts of the URL have to be specified in separate fields: `scheme`, `hostname`, and `pathprefix`
* The fields `scheme` and `pathprefix` are optional. If not specified, the OCM CLI will use the credentials for all schemes and paths on that host
* The `password` is the user's basic authentication password. Some OCI registries allow to generate user access tokens, which can also be used for basic authentication

```

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          scheme: https
          hostname: ghcr.io
          pathprefix: open-component-model
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-token
```

> #### HTTP, port number, empty path

Reviewer 4: ```suggestion
#### HTTP, Port Number, Empty Path

```

To access artifacts in `http://127.0.0.1:5001`:
* Note the quotes around the `port` number
* The fields `scheme` and  `port` are optional. If not specified, the OCM CLI will use the credentials for all schemes and ports on that host

> * As the URL has no path behind the port number, `pathprefix` element can be removed

Reviewer 4: ```suggestion
* As the URL has no path behind the port number, the `pathprefix` element can be removed

```

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          scheme: http
          hostname: 127.0.0.1
          port: "5001"
        credentials:
          - type: Credentials
            properties:
              username: admin
              password: admin
```

> ### Accessing Helm chart repositories

Reviewer 4: ```suggestion
### Accessing Helm Chart Repositories

```

Similar to OCI registries, but use `HelmChartRepository` as identity type.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: HelmChartRepository
          hostname: ghcr.io
          pathprefix: open-component-model
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-token
```

> ### Accessing Maven repositories

Reviewer 4: ```suggestion
### Accessing Maven Repositories

```

Similar to OCI registries, but use `MavenRepository` as identity type.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: MavenRepository
          hostname: maven.repo.host
          pathprefix: path/to/repo
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-password
```

> ### Accessing npm registries

Reviewer 4: ```suggestion
### Accessing npm Registries

```

Similar to OCI registries, but use `NpmRegistry` as identity type. In addition, it is required to specify the `email` address matching with the one in the user record in the npm registry.

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: NpmRegistry
          hostname: npm.registry.host
          pathprefix: path/to/registry
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-password
              email: foo.bar@acme.org
```

> ### Accessing GitHub repositories

Reviewer 4: ```suggestion
### Accessing GitHub Repositories

```

To access code in `https://my.github.enterprise/my-org/my-repo`:
* Use `Github` as identity type
* `hostname` is the domain name of the GitHub instance
* `pathprefix` is a combination of organization and repository names
* `token` is a personal access token generated in GitHub Developer Settings

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: Github
          hostname: my.github.enterprise
          pathprefix: my-org/my-repo
        credentials:
          - type: Credentials
            properties:
              token: ghp_my_personal_access_token
```

> ### Accessing several systems

Reviewer 4: ```suggestion
### Accessing Several Systems

```

configurations:
  - type: credentials.config.ocm.software
    repositories:
      - repository:
          type: NPMConfig/v1
          npmrcFile: '~/.npmrc'
```

### Accessing OCI Registries

#### HTTPS and Path

To access artifacts in `https://ghcr.io/open-component-model`:

* The different parts of the URL have to be specified in separate fields: `scheme`, `hostname`, and `pathprefix`
* The fields `scheme` and `pathprefix` are optional. If not specified, the OCM CLI will use the credentials for all schemes and paths on that host
* The `password` is the user's basic authentication password. Some OCI registries allow to generate user access tokens, which can also be used for basic authentication

```yaml
type: generic.config.ocm.software/v1
configurations:
  - type: credentials.config.ocm.software
    consumers:
      - identity:
          type: OCIRegistry
          scheme: https
          hostname: ghcr.io
          pathprefix: open-component-model
        credentials:
          - type: Credentials
            properties:
              username: some-user
              password: some-token
```

#### HTTP, Port Number, Empty Path

To access artifacts in `http://127.0.0.1:5001`:

> * Note the quotes around the `port` number

Reviewer 2: should be resolved by https://github.com/open-component-model/ocm/pull/927

Author: #927 works as expected. Quotes removed.
